### DISCUSSION TITLE
Help! Trying to "montage" without the white space.

### DISCUSSION THREAD

**Discussion by joshcali** [ImageMagick/ImageMagick#2475] [Help]
hi! I don't care about preserving aspect ratio, so I'm trying to create a montage but with all the images just turned into a square, so there's NO white space, and the images no longer match their original aspect ratio.

So basically instead of "fitting" an image into a square with white space, it basically fills the square completely by changing the shape of the image to be square, and the full size of the image

Is it possible in imagemagick? the montage command seems to just scale the images down and puts white space around them to preserve the aspect ratio.

thank you!

**Comment by GeeMack** [comment]
With ImageMagick's "montage" command you can use a "!" flag on your geometry HxW argument to force all images to some exact dimensions without regard to aspect ratio. And setting the +X+Y geometry to "+0+0" will append the images without any spacing.

Try this comand for example...

    montage *.jpg -geometry 128x128!+0+0 -tile x2 result.png

That reads in all the JPG images, resizes them all to 128 pixels square, and assembles them with no spacing to create the output. The "-tile x2" tells the command to make the grid as wide as needed to use all the images and two rows high.

**Reply by joshcali** [reply]
Thank you! So I'm on the mac, and when I try to add the ! into my command line, it gives me an error?

 josh$ magick geometry 512x512!+0+0 ./josh/p1_* images.jpg
-bash: !+0+0: event not found

is there a syntax to make the "!" work on osx?

**Reply by GeeMack** [reply]
I'm pretty sure with OS X you can just put those geometry arguments inside single quotes like...

     ... -geometry '512x512!+0+0' ...

**Reply by joshcali** [reply]
@GeeMack  thank you! hey, is there a way to always make the file a certain size (like 2048x2048) no matter how many tiles are full?

**Reply by GeeMack** [reply]
I don't use "montage" much, and didn't find a particular argument to resize the final output to certain dimensions, but it can be done pretty simply by sending the output of "montage" through a pipe and into "magick" like this...

    montage *.jpg -tile x2 -geometry '512x512!+0+0' MIFF:- | magick - -resize 1280x1280! final.png

That formats the output of "montage" as IM's proprietary MIFF format, sends it to standard out as "MIFF:-", pipes it into "magick" as "-", then resizes it and writes the output file.

For IMv6 use "convert" instead of "magick".

**Comment by fmw42** [comment]
To keep from having white space for equal size images, use -geometry +0+0. If you want equal size and no space use -geometry DxD+0+0! where D is the desired dimension with the exclamation to resize exactly with distortion. So for example:

`montage lena.jpg barn.jpg monet2.jpg zelda1.jpg -geometry 256x256+0+0! -tile 2x2 x.jpg`

![x](https://user-images.githubusercontent.com/13458218/91509518-50b18500-e88f-11ea-82d8-d11d40f2a6f8.jpg)

See
https://imagemagick.org/Usage/montage/#controls
https://imagemagick.org/script/command-line-processing.php#geometry

**Reply by fmw42** [reply]
Sorry, Greg. We seemed to have posted at the same time.

**Reply by joshcali** [reply]
hey @fmw42 thank you! 

is there a way to use the "!" in osx? every time I include it in a command line command it gives me back an error :)

**Comment by fmw42** [comment]
As GeeMack points out, best to put the values in (double) quotes. But you can also try escaping the ! as \!. So

`-geometry DxD+0+0\!`   or `-geometry "DxD+0+0!"`

**Reply by joshcali** [reply]
@fmw42  yes that worked! thank you! is there a way to always make the file a certain size (like 2048x2048) no matter how many tiles are full?

**Comment by fmw42** [comment]
You will have to compute the size for each tile given the number of tiles in x and y that would be needed to achieve your desired output size. Or after the montage, resize to the final size. If you use -geometry value for D as the minimum dimension of all your input images, then afterwards just resize to the desired output size. There is no automatic method that I know apart from resizing at the end.
